src/main/scala/contadamination/ContaminationModel.scala

```diff
@@ -0,0 +1,179 @@
+import org.apache.spark.rdd.RDD
+import org.bdgenomics.adam.rdd.ADAMContext
+import org.bdgenomics.formats.avro.AlignmentRecord
+
+import scala.collection
+
+object ContaminationModel {
+
+  val defaultWinSize = 32
+  val defaultFpr = 0.1
+  val defaultSeed = 0
+
+  case class FilterContext(
+                name: String,     // Contig name
+                path: String,     // Reference path
+                numEntries: Int,  // Number of sliding windows
+                fpr: Double,      // False Positive Rate
+                bloomFilter: BF            // Bloom filter
+  )
+
+  case class ContaminationStats(
+                numReads: Int,    // Total number of reads checked
+                tn: Int,          // True Negatives (number of rejected reads)
+                rate: Double      // Contamination rate as the number of rejected reads over the total number of reads
+  )
+  
+  /**
+   * Build a ContaminationModel from a sequence of paths
+   * @param paths Sequence of FASTA files for the reference contigs
+   * @param winSize Sliding window size
+   * @param fpr False Positive Rate
+   * @param adamContext ADAM Context
+   * @return a ContaminationModel
+   */
+  def apply(paths: Seq[String], winSize: Int = defaultWinSize, fpr: Double = defaultFpr)(implicit adamContext: ADAMContext): ContaminationModel = {
+    paths.map(ContaminationModel(_, winSize, fpr)).reduce((x, y) => x ++ y)
+  }
+  
+  /**
+   * Build a ContaminationModel from a path
+   * @param path FASTA file with the reference contig
+   * @param winSize Sliding window size
+   * @param fpr False Positive Rate for the BloomFilter
+   * @param adamContext ADAM Context
+   * @return a ContaminationModel
+   */
+  def apply(path: String, winSize: Int = defaultWinSize, fpr: Double = defaultFpr)(implicit adamContext: ADAMContext): ContaminationModel = {
+    val fragments = adamContext.loadFasta(path, winSize).cache()
+    require(fragments.count() > 0, "At least one fragment is required")
+
+    // TODO check that there is only one contigName per file (right now we assume there is only one contig per file)
+    val contigName = fragments.first().getContig.getContigName
+
+    val slidingFragments = new RDDFunctions(fragments).sliding(2).flatMap({
+      case Array() =>
+        Array()
+
+      case Array(frag) =>
+        val seq = frag.getFragmentSequence
+        for (i <- 0 until seq.size) yield seq.substring(i, winSize - i) + List.fill(i)(".").mkString
+
+      case Array(frag1, frag2) =>
+        val seq1 = frag1.getFragmentSequence
+        val seq2 = frag2.getFragmentSequence
+        val seq = seq1 + (if (seq2.size == seq1.size) seq2 else seq2 + List.fill(math.abs(seq1.size - seq2.size))(".").mkString)
+        for (i <- 0 until seq1.size) yield seq.substring(i, winSize)
+    }).cache()
+
+    // TODO I have concerns about the high number of entries and the user asking for a low FPR
+    val numEntries = slidingFragments.count().toInt
```

Comment:
I have this idea that maybe we could use (HyperLogLog)[https://en.wikipedia.org/wiki/HyperLogLog] to do approximate element counting as this should be accurate enough for this purpose. Accidentally there is a HyperLogLog implementation in Algebird. :smile: I'll check that out and see if it's something that we could use.